### PR TITLE
Add Theme Plugin support

### PR DESCRIPTION
**Pull request by kounoike**
This PR introduce "Theme Plugin" support. discoused in #1279.

Each GitBucket instance has some "theme"s. And each theme has some "skin"s.
ex.

- [AdminLTE theme](https://github.com/kounoike/gitbucket-theme-AdminLTE-plugin) is theme of AdminLTE theme.
    - skin-blue
    - skin-blue-light
    - skin-black
    - skin-black-light, etc.
- [AdminLTE pink theme](https://github.com/kounoike/gitbucket-theme-AdminLTE-pink-plugin) is demonstrate version theme.
    - skin-pink

In System settings, You can chose "skin name" with theme/skin pair.

140328-system settings - google chrome

first option is adminLTE-pink theme's skin. others are AdminLTE theme's skin.

theme/skin setting is saved into GITBUCKET_HOME/gitbucket.conf.

Every page of GitBucket load only selected theme's CSS/JS.

If you didn't have no theme-plugins, fallback to previous AdminLTE/skin-blue settings.

### Before submitting a pull-request to Gitbucket I have first:

- [x] read the [contribution guidelines](https://github.com/gitbucket/gitbucket/blob/master/.github/CONTRIBUTING.md)
- [x] rebased my branch over master
- [x] verified that project is compiling
- [x] verified that tests are passing
- [ ] squashed my commits as appropriate *(keep several commits if it is relevant to understand the PR)*
- [ ] [marked as closed using commit message](https://help.github.com/articles/closing-issues-via-commit-messages/) all issue ID that this PR should correct

### CODE REVIEW

File: src/main/twirl/gitbucket/core/main.scala.html [hunk 1]
```diff
@@ -15,10 +15,16 @@
...
-    <link href="@helpers.assets/vendors/AdminLTE-2.3.8/css/AdminLTE.min.css" rel="stylesheet">
-    <link href="@helpers.assets/vendors/AdminLTE-2.3.8/css/skins/skin-blue.min.css" rel="stylesheet">
     <link href="@helpers.assets/vendors/font-awesome-4.6.3/css/font-awesome.min.css" rel="stylesheet">
     <link href="@helpers.assets/common/css/gitbucket.css" rel="stylesheet">
+    @helpers.themePluginName.map { pluginName =>
+      @PluginRegistry().getThemeCssFiles(pluginName).map { file =>
+        <link href="@helpers.pluginAssets/@file" rel="stylesheet">
+      }
+    }.getOrElse{
+      <link href="@helpers.assets/vendors/AdminLTE-2.3.8/css/AdminLTE.min.css" rel="stylesheet">
```
Comment (takezoe) [inline src/main/twirl/gitbucket/core/main.scala.html]: I'm not convinced that `AdminLTE.min.css` should be replaceable by plugin.

File: src/main/scala/gitbucket/core/plugin/Plugin.scala [hunk 2]
```diff
@@ -190,6 +190,26 @@ abstract class Plugin {
   def suggestionProviders(registry: PluginRegistry, context: ServletContext, settings: SystemSettings): Seq[SuggestionProvider] = Nil
 
   /**
+    * Override and set true if this plugin is theme plugin.
+    */
+  val isThemePlugin: Boolean = false
```
Comment (takezoe) [inline src/main/scala/gitbucket/core/plugin/Plugin.scala]: Why this property is needed? It seems good to register overridden values of `themeCssFiles` and `themeSkinNames` to `PluginRegistry` simply.

**Comment by takezoe** [comment]
@kounoike Thanks for your great work!

As a whole, I felt that switching skin (of AdminLTE) is good, but replacing theme by plugin is not realistic. In fact, GitBucket HTML structure depends on AdminLTE strongly. Do you think it's possible to modify total design by replacing AdminLTE's CSS?

**Comment by kounoike** [comment]
@takezoe Thanks for comments.

Yes, it is hard work to completely replace AdminLTE. But, theme developer can **modify** it.
Now I'm trying change gitbucket-theme-AdminLTE-plugin to building CSS from LESS with sbt-less.
It can easy to modify AdminLTE's layout such as sidebar-width, etc.

I think move gitbucket.css to plugin is more useful for theme developer. What do you think?

File: src/main/scala/gitbucket/core/plugin/Plugin.scala (hunk 2 above)
Comment (kounoike) [inline src/main/scala/gitbucket/core/plugin/Plugin.scala]: OK. I will remove it.

File: src/main/scala/gitbucket/core/plugin/PluginRegistory.scala [hunk 3]
```diff
@@ -146,6 +150,19 @@ class PluginRegistry {
   def addSuggestionProvider(suggestionProvider: SuggestionProvider): Unit = suggestionProviders += suggestionProvider
 
   def getSuggestionProviders: Seq[SuggestionProvider] = suggestionProviders.toSeq
+
+  def addThemeJavaScriptFile(pluginId: String, file: String): Unit = themeJavaScriptFiles += ((pluginId, file))
+
+  // def getThemeJavaScriptFiles(pluginId): List[String] = themeJavaScriptFiles.filter(x => pluginId.matches(x._1)).toList.map(_._2)
```
Comment (takezoe) [inline src/main/scala/gitbucket/core/plugin/PluginRegistory.scala]: You should delete this comment line.

**Comment by takezoe** [comment]
> Yes, it is hard work to completely replace AdminLTE. But, theme developer can modify it.

I can't decide we should continue to use AdminLTE in the future. If we add extension points that depend on AdminLTE, it will be difficult to remove AdminLTE in the future. In this point of view, the skin is also AdminLTE specific feature. We have to carefully consider to add new extension points in this case.

> I think move gitbucket.css to plugin is more useful for theme developer.

Capability of user defined CSS and JavaScript might be a solution.

**Comment by kounoike** [comment]
I thought few days, maybe it can make alternative CSS, but maybe no user do it.
Shall I make new PR? which replaces skin only.

**Comment by takezoe** [comment]
@kounoike Sorry, I haven't enough considered about this topic. Could you give me some more time?

**Comment by kounoike** [comment]
@takezoe OK. Once you have decided on the policy, I will support this topic.

**Pull request closed without merging**